cookbooks/prescribed-velocity/prescribed_velocity.cc

```diff
@@ -0,0 +1,200 @@
+                // If it's okay to constrain this DOF
+                if (current_constraints.can_store_line(local_dof_indices[q]) &&
+                    !current_constraints.is_constrained(local_dof_indices[q]))
+                  {
+                    const unsigned int component_index = simulator_access.get_fe().system_to_component_index(q).first;
+
+                    // If we're on one of the velocity DOFs
+                    if (component_index >= simulator_access.introspection().component_indices.velocities[0] &&
+                        component_index <= simulator_access.introspection().component_indices.velocities[dim-1])
+                      {
+                        // Which velocity component is this DOF associated with?
+                        const unsigned int component_direction = component_index
+                                                                 - simulator_access.introspection().component_indices.velocities[0];
+
+                        // we get time passed as seconds (always) but may want
+                        // to reinterpret it in years
+                        double time = simulator_access.get_time();
+                        if (simulator_access.convert_output_to_years())
+                          time /= (60*60*24*365.2425);
+
+                        prescribed_velocity_indicator_function_2d.set_time (time);
+                        prescribed_velocity_indicator_function_3d.set_time (time);
+                        prescribed_velocity_function_2d.set_time (time);
+                        prescribed_velocity_function_3d.set_time (time);
+
+                        // Awkward because we need cases for 2d and 3d
+                        Point<dim> p = fe_values.quadrature_point(q);
+                        Point<2> p2d = Point<2> (p[0], p[1]);
+                        Point<3> p3d = Point<3> (p[0], p[1], (dim == 3) ? p[2] : 0);
+                        double indicator, u_i;
+                        if (dim == 2)
+                          {
+                            indicator = prescribed_velocity_indicator_function_2d.value (p2d, component_direction);
+                            u_i = prescribed_velocity_function_2d.value (p2d, component_direction);
+                          }
+                        else
+                          {
+                            indicator = prescribed_velocity_indicator_function_3d.value (p3d, component_direction);
+                            u_i = prescribed_velocity_function_3d.value (p3d, component_direction);
+                          }
+
+                        if (indicator > 0)
+                          {
+                            // Add line to mass matrix (if there isn't already one for this DOF)
+                            current_constraints.add_line (local_dof_indices[q]);
+                            // Add constraint to this DOF
+                            current_constraints.set_inhomogeneity (local_dof_indices[q], u_i);
```

Comment:
it's not the mass matrix but the constraint matrix. I would write the comment for these two lines of code as
```
// Add a constraint of the form  dof[q] = u_i  to the list of constraints.
```